lib/models/event-timeline.js

```diff
@@ -0,0 +1,252 @@
+"use strict";
+
+/**
+ * @module models/event-timeline
+ */
+
+var RoomState = require("./room-state");
+
+/**
+ * Construct a new EventTimeline
+ *
+ * <p>An EventTimeline represents a contiguous sequence of events in a room.
+ *
+ * <p>As well as keeping track of the events themselves, it stores the state of
+ * the room at the beginning and end of the timeline, and pagination tokens for
+ * going backwards and forwards in the timeline.
+ *
+ * <p>In order that clients can meaningfully maintain an index into a timeline, we
+ * track a 'baseIndex'. This starts at zero, but is incremented when events are
+ * prepended to the timeline. The index of an event relative to baseIndex
+ * therefore remains constant.
+ *
+ * <p>Once a timeline joins up with its neighbour, we link them together into a
+ * doubly-linked list.
+ *
+ * @param {string} roomId    the ID of the room where this timeline came from
+ * @constructor
+ */
+function EventTimeline(roomId) {
+    this._roomId = roomId;
+    this._events = [];
+    this._baseIndex = -1;
+    this._startState = new RoomState(roomId);
+    this._endState = new RoomState(roomId);
+
+    this._prevTimeline = null;
+    this._nextTimeline = null;
+
+    // this is used by client.js
+    this._paginationRequests = {'b': null, 'f': null};
+}
+
+/**
+ * Initialise the start and end state with the given events
+ *
+ * <p>This can only be called before any events are added.
+ *
+ * @param {MatrixEvent[]} stateEvents list of state events to intialise the
+ * state with.
+ */
+EventTimeline.prototype.initialiseState = function(stateEvents) {
+    if (this._events.length > 0) {
+        throw new Error("Cannot initialise state after events are added");
+    }
+
+    // do we need to copy here? sync thinks we do but I can't see why
+    this._startState.setStateEvents(stateEvents);
+    this._endState.setStateEvents(stateEvents);
```

Comment:
You need to deep copy here because otherwise these `_startState` and `_endState` will be backed by the *same* object and the *same* events, so mutations in the object/random state events will affect *both* old and current state. This can introduce subtle bugs.